### PR TITLE
gh-128192: support HTTP sha-256 digest authentication as per RFC-7617

### PR DESCRIPTION
**Pull request by calvinbui**
as mentioned in the issue, other authentication exist, but i don't see them supported out of the box in hashlib.

this all depends if python wishes to support rfc7616

* Issue: gh-128192

### CODE REVIEW

File: Misc/NEWS.d/next/Core_and_Builtins/2024-12-23-11-14-07.gh-issue-128192.02mEhD.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Support digest authentication algorithm SHA-256
```
Comment (ZeroIntensity) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-12-23-11-14-07.gh-issue-128192.02mEhD.rst]: We need context:
```suggestion
Support digest authentication algorithm SHA-256 in :mod:`urllib.request`.
```

**Comment by ZeroIntensity** [comment]
cc @picnixz

File: Misc/NEWS.d/next/Core_and_Builtins/2024-12-23-11-14-07.gh-issue-128192.02mEhD.rst (hunk 1 above)
Comment (calvinbui) [reply Misc/NEWS.d/next/Core_and_Builtins/2024-12-23-11-14-07.gh-issue-128192.02mEhD.rst]: thanks!

File: Lib/urllib/request.py
Comment (rruuaanng) [inline Lib/urllib/request.py:1]: It looks fine, but you should add a test.

from RFC-7616
```
   This specification defines the following algorithms:

   o  SHA2-256 (mandatory to implement)

   o  SHA2-512/256 (as a backup algorithm)

   o  MD5 (for backward compatibility).
```

Comment (calvinbui) [reply Lib/urllib/request.py:1]: i dont know, I assume i would just be re-writing a test that probably exists in hashlib?

i've only found a [single test](https://github.com/python/cpython/blob/3ddd70ceaaf67b111ee4251817e150396d6d10a9/Lib/test/test_urllib2.py#L1965) written for this function that checks for an invalid algorithm.

Comment (picnixz) [reply Lib/urllib/request.py:1]: `hashlib` does not need to be tested. What needs to be tested is that `MD5`, `SHA` and `SHA-256` are recognized algorithms if you specify them as is in the request headers. This is an opportunity to add tests for this as well.

File: Lib/urllib/request.py [hunk 2]
```diff
@@ -1182,6 +1182,8 @@ def get_algorithm_impls(self, algorithm):
         elif algorithm == 'SHA':
             H = lambda x: hashlib.sha1(x.encode("ascii")).hexdigest()
         # XXX MD5-sess
+        elif algorithm == 'SHA-256':
```
Comment (picnixz) [inline Lib/urllib/request.py:1185]: can you move the `XXX MD5-sess` before the else and not before the `elif`? thanks. In addition, mention that the algorithm names are taken from https://datatracker.ietf.org/doc/html/rfc7616#section-6.1.

File: Misc/NEWS.d/next/Core_and_Builtins/2024-12-23-11-14-07.gh-issue-128192.02mEhD.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Support digest authentication algorithm SHA-256 in :mod:`urllib.request`.
```
Comment (picnixz) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-12-23-11-14-07.gh-issue-128192.02mEhD.rst]:
```suggestion
Upgrade HTTP digest authentication algorithm for :mod:`urllib.request` by
supporting SHA-256 digest authentication as specified in :rfc:`7616`.
```

This also requires a What's New entry. However, we do not have any documentation concerning HTTP digest access authentication (we only have https://docs.python.org/3/library/urllib.request.html#abstractdigestauthhandler-objects but never do we mention which digests we are supporting). Nonetheless, maybe we can update the wording of that documented entry.

**Review by picnixz — COMMENTED** [review]
A preliminary round of comments. You can also update "RFC 2617" to "RFC 2617/7616" in the AbstractDigestAuthHandler comment.

File: Misc/NEWS.d/next/Core_and_Builtins/2024-12-23-11-14-07.gh-issue-128192.02mEhD.rst (hunk 3 above)
Comment (calvinbui) [reply Misc/NEWS.d/next/Core_and_Builtins/2024-12-23-11-14-07.gh-issue-128192.02mEhD.rst]: thanks!

File: Lib/urllib/request.py (hunk 2 above)
Comment (calvinbui) [reply Lib/urllib/request.py:1185]: done:

- https://github.com/python/cpython/pull/128193/commits/5c214022d6aff41069b65a587cd2632777e158cd
- https://github.com/python/cpython/pull/128193/commits/aea1b4ec777ddde9cc8ecfdc99e4f3762313c08d

File: Lib/urllib/request.py
Comment (calvinbui) [reply Lib/urllib/request.py:1]: tests added

File: Lib/test/test_urllib2.py [hunk 4]
```diff
@@ -1962,10 +1962,28 @@ def test_parse_proxy(self):
 
         self.assertRaises(ValueError, _parse_proxy, 'file:/ftp.example.com'),
 
-    def test_unsupported_algorithm(self):
-        handler = AbstractDigestAuthHandler()
+class TestDigestAlgorithms(unittest.TestCase):
```
Comment (picnixz) [inline Lib/test/test_urllib2.py:1966]:
```suggestion

class TestDigestAlgorithms(unittest.TestCase):
```

File: Lib/test/test_urllib2.py [hunk 5]
```diff
@@ -1962,10 +1962,28 @@ def test_parse_proxy(self):
...
+    def setUp(self):
+        self.handler = AbstractDigestAuthHandler()
```
Comment (picnixz) [inline Lib/test/test_urllib2.py:1968]: In a follow-up PR (or this one), if you want/can, we can add tests for a full communication round where we request HTTP digest authentication.

Comment (calvinbui) [reply Lib/test/test_urllib2.py:1968]: yep, next time!

**Comment by picnixz** [comment]
We are still missing the What's New entry and the modified NEWS entry.

**Comment by calvinbui** [comment]
> We are still missing the What's New entry and the modified NEWS entry.

my bad, fixed

File: Doc/whatsnew/3.14.rst [hunk 6]
```diff
@@ -638,6 +638,13 @@ unittest
   (Contributed by Jacob Walls in :gh:`80958`.)
 
 
+urllib
+------
+
+* Upgrade HTTP digest authentication algorithm for :mod:`urllib.request` by
+  supporting SHA-256 digest authentication as specified in :rfc:`7616`.
+  (Contributed by Calvin Bui in :gh:`128193`)
+
```
Comment (picnixz) [inline Doc/whatsnew/3.14.rst]:
```suggestion
  (Contributed by Calvin Bui in :gh:`128193`.)


```

**Review by picnixz — COMMENTED** [review]
We may also need an additional `.. versionchanged:: next` in the docs to mention SHA-256 in https://docs.python.org/3/library/urllib.request.html#urllib.request.AbstractDigestAuthHandler.

I think we'll also need a follow-up PR to update the docs because they are lacking IMO.

**Comment by calvinbui** [comment]
> We may also need an additional `.. versionchanged:: next` in the docs to mention SHA-256 in https://docs.python.org/3/library/urllib.request.html#urllib.request.AbstractDigestAuthHandler.

added
 
> I think we'll also need a follow-up PR to update the docs because they are lacking IMO.

what docs were you thinking of? i can make a new gh issue for it, as well as the complete end-to-end http digest auth test

**Comment by picnixz** [comment]
> what docs were you thinking of? i can make a new gh issue for it, as well as the complete end-to-end http digest auth test

None of the methods of AbstractDigestAuthHandler are actually documented so we may want to document them. If someone wants to subclass this interface, they need to know its usage.

File: Lib/urllib/request.py [hunk 7]
```diff
@@ -1176,11 +1176,14 @@ def get_authorization(self, req, chal):
         return base
 
     def get_algorithm_impls(self, algorithm):
+        # algorithm names taken from RFC 7616 Section 6.1
         # lambdas assume digest modules are imported at the top level
         if algorithm == 'MD5':
             H = lambda x: hashlib.md5(x.encode("ascii")).hexdigest()
         elif algorithm == 'SHA':
```
Comment (gpshead) [inline Lib/urllib/request.py]:
```suggestion
        elif algorithm == 'SHA':  # non-standard, retained for compatibility.
```

**Review by gpshead — APPROVED** [review]

**Comment by picnixz** [comment]
Thank you for your contribution Calvin!

**Comment by gpshead** [comment]
i'm not surprised at that buildbot error.  it is running in an unrealistic config without critical hash functions.  we'll ultimately just need to decorate the test to indicate that it requires md5, sha1, and sha256 - skipping it otherwise.  (see test_hashlib)

**Comment by picnixz** [comment]
I'll do it tomorrow (it's almost midnight here and I'm no more on my dev session) or someone else somewhere else can add the decorators. However, maybe we should revert the commit just to avoid other PRs to have the buildbot failure in the meantime?

**Comment by gpshead** [comment]
No worries. That buildbot is an unstable unsupported configuration, no need to revert anything.

It's there solely so that we can use it to identify and clean up test dependency declaration issues _eventually_. It doesn't block anything.

**Pull request merged**